### PR TITLE
Replace custom theme with sphinx-ansible-theme

### PR DESCRIPTION
**Pull request by ssbarnea**
- remove custom sphinx theme from our codebase
- reuse shared sphinx-ansible-theme
- removed docs makefiles

### CODE REVIEW

File: docs/docsite/rst/conf.py [hunk 1]
```diff
@@ -125,9 +128,22 @@
...
-html_theme_path = ['../_themes']
-html_theme = 'sphinx_rtd_theme'
-html_short_title = 'Ansible Documentation'
+html_theme = 'sphinx_ansible_theme'
+
+
+html_context = {
+    'display_github': 'True',
+    'github_user': 'ansible',
+    'github_repo': 'ansiblelint',
```
Comment (webknjaz) [inline docs/docsite/rst/conf.py:137]: typo?

File: docs/docsite/rst/conf.py [hunk 2]
```diff
@@ -193,6 +210,19 @@
 # Output file base name for HTML help builder.
 htmlhelp_basename = 'Poseidodoc'
 
+# Configuration for sphinx-notfound-pages
+# with no 'notfound_template' and no 'notfound_context' set,
+# the extension builds 404.rst into a location-agnostic 404 page
+#
+# default is `en` - using this for the sub-site:
+notfound_default_language = "ansible"
```
Comment (webknjaz) [inline docs/docsite/rst/conf.py:218]: What is this?

File: docs/docsite/rst/conf.py [hunk 3]
```diff
@@ -235,9 +265,3 @@
 
 
 # table width fix via: https://rackerlabs.github.io/docs-rackspace/tools/rtd-tables.html
```
Comment (webknjaz) [inline docs/docsite/rst/conf.py:267]: This refers to a deleted block

File: lib/ansiblelint/__init__.py [hunk 4]
```diff
@@ -19,5 +19,12 @@
 # THE SOFTWARE.
 """Main ansible-lint package."""
 
+try:
+    import pkg_resources
```
Comment (webknjaz) [inline lib/ansiblelint/__init__.py:23]: Don't add more than one instruction in try/except

File: lib/ansiblelint/__init__.py [hunk 5]
```diff
@@ -19,5 +19,12 @@
...
+
+    __version__ = pkg_resources.get_distribution(__package__).version
+except Exception:
```
Comment (webknjaz) [inline lib/ansiblelint/__init__.py:26]: Don't use catch-all exceptions.

File: README.rst [hunk 6]
```diff
@@ -497,8 +497,8 @@ It's also a good practice to comment the reasons why a task is being skipped.
 
 If you want skip running a rule entirely, you can use either:
 
-* `command-line skip_list`_ via ``-x``
-* `config file skip_list`_
+* ``command-line skip_list`` via ``-x``
```
Comment (webknjaz) [inline README.rst:500]: out of scope

File: lib/ansiblelint/__init__.py [hunk 7]
```diff
@@ -19,5 +19,12 @@
...
+    __version__ = "unknown"
+
```
Comment (webknjaz) [inline lib/ansiblelint/__init__.py:28]: out of scope

File: setup.cfg [hunk 8]
```diff
@@ -77,6 +77,11 @@ install_requires =
   # NOTE: per issue #509 0.15.34 included in debian backports
   typing-extensions; python_version < "3.8"
 
+[options.extras_require]
+docs =
+    Sphinx
+    sphinx_ansible_theme >= 0.2.2
+
```
Comment (webknjaz) [inline setup.cfg:84]: Extras are for *runtime* deps what enable the project do additional stuff when you run it. The right way is to use https://github.com/ansible/ansible-lint/pull/861 (which by itself belongs to a separate PR).

File: tox.ini [hunk 9]
```diff
@@ -103,9 +103,25 @@ passenv =
...
+      -b html --color \
+      -d "{toxinidir}/docs/docstree" \
+      docs/docsite/rst/ "{toxinidir}/docs/docstree/html"
+
+    # Print out the output docs dir and a way to serve html:
+    python -c \
+    'import pathlib; '\
+    'docs_dir = pathlib.Path(r"{toxinidir}") / "docs/html"; index_file = docs_dir / "index.html"; print(f"\nDocumentation available under `file://\{index_file\}`\n\nTo serve docs, use `python3 -m http.server --directory \{docs_dir\} 0`\n")'
+extras =
+    docs
```
Comment (webknjaz) [inline tox.ini:124]: out of scope

File: docs/docsite/Makefile.sphinx [hunk 10]
```diff
@@ -1,25 +0,0 @@
-# Minimal makefile for Sphinx documentation
```
Comment (webknjaz) [inline docs/docsite/Makefile.sphinx:1]: out of scope

File: docs/docsite/Makefile [hunk 11]
```diff
@@ -1,75 +0,0 @@
-OS := $(shell uname -s)
```
Comment (webknjaz) [inline docs/docsite/Makefile:1]: out of scope

File: README.rst [hunk 12]
```diff
@@ -597,5 +597,3 @@ ansible-lint was created by `Will Thames`_ and is now maintained as part of the
 .. _Will Thames: https://github.com/willthames
 .. _Ansible: https://ansible.com
 .. _Red Hat: https://redhat.com
-.. _command-line skip_list: https://docs.ansible.com/ansible-lint/usage/usage.html#command-line-options
-.. _config file skip_list: https://docs.ansible.com/ansible-lint/configuring/configuring.html#configuration-file
```
Comment (webknjaz) [inline README.rst:601]: out of scope

File: .gitignore [hunk 13]
```diff
@@ -41,3 +41,4 @@ pip-wheel-metadata
 
 # mypy
 .mypy_cache
+docs/docstree
```
Comment (webknjaz) [inline .gitignore:44]: out of scope. also invalid location.

File: .readthedocs.yml [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+build:
+  image: latest
+
+python:
+  version: 3.7
+  install:
+    - method: pip
+      path: .
+      extra_requirements:
+        - docs
```
Comment (webknjaz) [inline .readthedocs.yml:15]: out of scope

File: docs/docsite/rst/conf.py [hunk 15]
```diff
@@ -193,6 +210,19 @@
...
+#
+# default is `en` - using this for the sub-site:
+notfound_default_language = "ansible"
+# default is `latest`:
+# setting explicitly - docsite serves up /ansible/latest/404.html
+# so keep this set to `latest` even on the `devel` branch
+# then no maintenance is needed when we branch a new stable_x.x
+notfound_default_version = "latest"
+# makes default setting explicit:
+notfound_no_urls_prefix = False
```
Comment (webknjaz) [inline docs/docsite/rst/conf.py:225]: doesn't look related

File: docs/docsite/rst/conf.py [hunk 16]
```diff
@@ -56,8 +59,8 @@
 master_doc = 'index'
 
 # General substitutions.
-project = 'Ansible Documentation'
-copyright = "2013-2018 Ansible, Inc"
+project = 'AnsibleLint'
+copyright = "2020 Ansible, Inc"
```
Comment (webknjaz) [inline docs/docsite/rst/conf.py:63]: unrelated in adding a theme

File: docs/docsite/rst/conf.py [hunk 17]
```diff
@@ -125,9 +128,22 @@
...
+
+
+html_context = {
+    'display_github': 'True',
+    'github_user': 'ansible',
+    'github_repo': 'ansiblelint',
+    'github_version': 'master/docs/',
+    'current_version': version,
+    'latest_version': 'latest',
+    # list specifically out of order to make latest work
```
Comment (webknjaz) [inline docs/docsite/rst/conf.py:141]: ?

File: docs/docsite/rst/conf.py [hunk 18]
```diff
@@ -125,9 +128,22 @@
...
 # -----------------------
 
-html_theme_path = ['../_themes']
-html_theme = 'sphinx_rtd_theme'
-html_short_title = 'Ansible Documentation'
+html_theme = 'sphinx_ansible_theme'
+
+
+html_context = {
+    'display_github': 'True',
```
Comment (webknjaz) [inline docs/docsite/rst/conf.py:135]: str/bool?

File: README.rst (hunk 12 above)
Comment (ssbarnea) [reply README.rst:601]: not really, I was forced to touch these two links because it was failing to build with them. I am not yet sure what confused sphinx about them. Anyway, the second issue of links to wrong docs location is also sorted. Yep, i can split it, but it needs to be addressed before we merge this.

File: .flake8 [hunk 19]
```diff
@@ -66,6 +66,8 @@ per-file-ignores =
   # FIXME: drop PT009 once they're fixes
   # Ref: https://github.com/ansible/ansible-lint/issues/725
   test/*.py: PT009 D100 D101 D102 D103
+  # Docs config is not public
+  docs/docsite/rst/conf.py: D
```
Comment (webknjaz) [inline .flake8:70]: Let's just not add violations there.

File: .readthedocs.yml [hunk 20]
```diff
@@ -0,0 +1,14 @@
...
+
+build:
+  image: latest
+
+python:
+  version: 3.7
+  install:
+    - requirements: docs/requirements.txt
+    - method: pip
+      path: .
```
Comment (webknjaz) [inline .readthedocs.yml:14]: still out of scope

File: docs/docsite/rst/conf.py [hunk 21]
```diff
@@ -12,34 +12,26 @@
 #
 # All configuration values have a default value; values that are commented out
 # serve to show the default value.
-"""Documentation Configuration."""
 
-import sys
-import os
+import datetime
```
Comment (webknjaz) [inline docs/docsite/rst/conf.py:16]: Refactoring this module is out of scope. The change should only contain the integration of the new theme.

File: docs/requirements.txt [hunk 22]
```diff
@@ -0,0 +1,2 @@
+Sphinx>=3.1.1
```
Comment (webknjaz) [inline docs/requirements.txt:1]: Out of scope and is addressed by https://github.com/ansible/ansible-lint/pull/861

File: lib/ansiblelint/__init__.py [hunk 23]
```diff
@@ -19,5 +19,11 @@
...
-# deprecated backwards compability:
-from ansiblelint.rules import AnsibleLintRule  # noqa: F401 api backwards compability
+from ansiblelint.rules import AnsibleLintRule
+from ansiblelint.version import __version__
+
+
+__all__ = [
+    "__version__",
+    "AnsibleLintRule"  # deprecated, import it directly from rules
+]
```
Comment (webknjaz) [inline lib/ansiblelint/__init__.py:29]: out of scope

File: tox.ini [hunk 24]
```diff
@@ -103,9 +103,22 @@ passenv =
...
+    python -m sphinx \
+      -a -n -W \
+      -b html --color \
+      -d "{toxworkdir}/.docstree" \
+      docs/docsite/rst/ "{toxinidir}/docs/html"
+
+    # Print out the output docs dir and a way to serve html:
+    python -c \
+    'import pathlib; '\
+    'docs_dir = pathlib.Path(r"{toxinidir}") / "docs/html"; index_file = docs_dir / "index.html"; print(f"\nDocumentation available under `file://\{index_file\}`\n\nTo serve docs, use `python3 -m http.server --directory \{docs_dir\} 0`\n")'
```
Comment (webknjaz) [inline tox.ini:121]: should be in https://github.com/ansible/ansible-lint/pull/858

File: docs/docsite/rst/conf.py (hunk 2 above)
Comment (ssbarnea) [reply docs/docsite/rst/conf.py]: Got it from Ansible itself, a guess is default code-block when language is not mentioned.

File: .readthedocs.yml [hunk 25]
```diff
@@ -36,5 +36,7 @@ build:
 python:
   version: 3.8
   install:
+  - method: pip
```
Comment (webknjaz) [inline .readthedocs.yml:39]: Let's add this to `docs/requirements.in` instead.

**Comment by webknjaz** [comment]
> docs/_themes/sphinx_rtd_theme/static/images/logo_invert.png

@ssbarnea should we keep this?

**Comment by ssbarnea** [comment]
Feel free to make any changes you want to the conf.py, i will not mind as long it builds with them.

File: docs/docsite/README.md [hunk 26]
```diff
@@ -0,0 +1,28 @@
+# Documentation source for Ansible Lint
```
Comment (webknjaz) [inline docs/docsite/README.md:1]: @ssbarnea this should be moved out of `docsite/` that doesn't exist anymore.

File: docs/rst/conf.py [hunk 27]
```diff
@@ -1,4 +1,5 @@
 # -*- coding: utf-8 -*-
+"""Sphinx configuration."""
```
Comment (webknjaz) [inline docs/rst/conf.py:2]: I don't think it should be moved here. Also it doesn't seem related to changing the theme anyway.

File: docs/rst/conf.py [hunk 28]
```diff
@@ -12,34 +13,26 @@
...
-"""Documentation Configuration."""
 
-import sys
-import os
+import datetime
 
-# pip install sphinx_rtd_theme
-# import sphinx_rtd_theme
-# html_theme_path = [sphinx_rtd_theme.get_html_theme_path()]
+# pip install sphinx_ansible_theme
```
Comment (webknjaz) [inline docs/rst/conf.py:19]: I don't think this is useful since all the supported configs have this info in the right place
```suggestion
```

File: docs/rst/conf.py [hunk 29]
```diff
@@ -12,34 +13,26 @@
...
+import sphinx_ansible_theme
+from ansiblelint import __version__, __package__
```
Comment (webknjaz) [inline docs/rst/conf.py:21]: Setting up version vars is out of scope and should be in a separate PR.

File: docs/rst/conf.py [hunk 30]
```diff
@@ -12,34 +13,26 @@
...
+# If your documentation needs a minimal Sphinx version, state it here.
+# needs_sphinx = '1.0'
 
-# Add any Sphinx extension module names here, as strings.
-# They can be extensions
-# coming with Sphinx (named 'sphinx.ext.*') or your custom ones.
-# TEST: 'sphinxcontrib.fulltoc'
+# Add any Sphinx extension module names here, as strings. They can be
+# extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
+# ones.
```
Comment (webknjaz) [inline docs/rst/conf.py:35]: this is not related

File: docs/README.md [hunk 31]
```diff
@@ -1,9 +1,28 @@
 # Documentation source for Ansible Lint
 
-To build the docs, run `tox -e docs`. At the end of the build, you will
-see the local location of your built docs.
+Contains the source behind [docs.ansible.com/ansible-lint](https://docs.ansible.com/ansible-lint)
 
-Building docs locally may not be identical to CI/CD builds. We recommend
-you to create a draft PR and check the RTD PR preview page too.
```
Comment (webknjaz) [inline docs/README.md]: @ssbarnea bad merge?

File: .readthedocs.yml (hunk 25 above)
Comment (ssbarnea) [reply .readthedocs.yml:39]: I am confident that this may introduce problems to tox, dependabot, and possible even RTD as it would require a relative path ".." inside a reqs file. Lets keep it out so we can control how we install our package opposed to the docs requirements. Even RTD examples display that.

File: docs/requirements.in [hunk 32]
```diff
@@ -1 +1,2 @@
 Sphinx
+sphinx_ansible_theme
```
Comment (webknjaz) [inline docs/requirements.in:2]: as per earlier suggestion
```suggestion
sphinx_ansible_theme
..
```

File: .readthedocs.yml (hunk 25 above)
Comment (webknjaz) [reply .readthedocs.yml:39]: @ssbarnea ^

File: docs/rst/conf.py [hunk 33]
```diff
@@ -126,7 +126,7 @@
 # -----------------------
 
 html_theme_path = ['../_themes']
-html_theme = 'sphinx_rtd_theme'
+html_theme = 'sphinx_ansible_theme'
```
Comment (webknjaz) [inline docs/rst/conf.py:129]: :+1:

File: docs/requirements.in (hunk 32 above)
Comment (ssbarnea) [reply docs/requirements.in:2]: Commented above. We can try this approach after enabling dependabot, I want to be sure it works well with it too.

**Comment by ssbarnea** [comment]
> Please answer this question:
> 
> > > docs/_themes/sphinx_rtd_theme/static/images/logo_invert.png
> > 
> > 
> > @ssbarnea should we keep this?
> 
> Also, for any change you want to do as a follow-up I expect a comment and maybe a link to an issue documenting that this needs to be done.

That is Ansible owns logo, now included with the theme. No need to keep it.
